### PR TITLE
HPCC-24827 Support fpos batch streaming in dafilesrv.

### PR DESCRIPTION
**Pull request by jakesmith**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday @rpastrana - please review.

NB: the changes in dafsstream.* and wsdfuaccess, are only used by unittests (and datest), and will soon be deprecated once, but I updated, as it's the easiest way I can test/add unittests at the moment.

File: fs/dafsstream/dafsstream.cpp
Comment (ghalliday) [inline fs/dafsstream/dafsstream.cpp:812]: Will this overwrite the list of fetch positions returned in the continuation record, if all the fpos haven't been matched?

File: system/jlib/jptree.cpp
Comment (ghalliday) [inline system/jlib/jptree.cpp:7286]: can rebase now.

**Review by ghalliday — COMMENTED** [review]
@jakesmith a few minor comments, 1 question.

File: fs/dafsstream/dafsstream.cpp
Comment (rpastrana) [inline fs/dafsstream/dafsstream.cpp:793]: very minor, but there seems to be a whitespace issue here

**Review by rpastrana — COMMENTED** [review]
@jakesmith a couple of minor comments.

File: fs/dafsserver/dafsserver.cpp
Comment (jakesmith) [reply fs/dafsserver/dafsserver.cpp:1257]: I'd favour readibility over size saving here, as never going to be many of these objects,
but it's also more readable if bools together, I'll move it.

File: fs/dafsserver/dafsserver.cpp
Comment (jakesmith) [reply fs/dafsserver/dafsserver.cpp:2134]: yes, will add.

File: fs/dafsstream/dafsstream.cpp
Comment (jakesmith) [reply fs/dafsstream/dafsstream.cpp:812]: It will add them at the server, but it does highlight a problem.
If reply limit hit, and therefore unconsumed fpos' at server, this will get another 'max' and send.
That means the amount will grow each round.

Need to check amount consumed and send max-consumed.
The cursor being sent back in a fetch has this info, at the moment, the client is agnostic to what the cursor contains, but I'll probably parse it for this info. when fetching.

File: fs/dafsserver/dafsserver.cpp
Comment (jakesmith) [reply fs/dafsserver/dafsserver.cpp:2145]: yes, will add.

File: fs/dafsserver/dafsserver.cpp
Comment (jakesmith) [reply fs/dafsserver/dafsserver.cpp:2151]: I think I'll leave as is, it's clear when constant that represents an xpath as it does in these cases (and other existing use cases), and would be less clear when using macros to build up xpaths from >1 macro e.g. queryProp(NODESTR "/" FILENAMESTR) vs queryProp("node/fileName")

File: fs/dafsstream/dafsstream.cpp
Comment (jakesmith) [reply fs/dafsstream/dafsstream.cpp:793]: there was a trailing white space, will remove.

**Comment by jakesmith** [comment]
@ghalliday @rpastrana - please see new commit.

**Review by rpastrana — APPROVED** [review]
@jakesmith approved. 
I did mean to ask, based on your in-code comment,files are not closed after a fetch is processed. What are the drawbacks of this behavior? and do we ever end up "closing" those files?

**Comment by jakesmith** [comment]
>I did mean to ask, based on your in-code comment,files are not closed after a fetch is processed. What are the drawbacks of this behavior? and do we ever end up "closing" those files?

Normally, a file will be closed by the client, or if the client socket goes away.
The close on streaming, at end of stream, is just a [small] optimization really, as the file is no longer of any use.
But in the fetch case that's not the case.

**Review by ghalliday — APPROVED** [review]
Looks good.  Will merge once it is merged.

**Pull request merged**
